src/diagram.rs

```diff
@@ -0,0 +1,669 @@
+    ///         }
+    ///     }
+    /// }
+    /// # "#)?;
+    /// # Ok::<_, serde_json::Error>(())
+    /// ```
+    Split(SplitOp),
+
+    /// Wait for an item to be emitted from each of the inputs, then combined the
+    /// oldest of each into an array.
+    ///
+    /// # Examples
+    /// ```
+    /// # bevy_impulse::Diagram::from_json_str(r#"
+    /// {
+    ///     "start": "split",
+    ///     "ops": {
+    ///         "split": {
+    ///             "type": "split",
+    ///             "index": ["op1", "op2"]
+    ///         },
+    ///         "op1": {
+    ///             "type": "node",
+    ///             "builder": "foo",
+    ///             "next": "join"
+    ///         },
+    ///         "op2": {
+    ///             "type": "node",
+    ///             "builder": "bar",
+    ///             "next": "join"
+    ///         },
+    ///         "join": {
+    ///             "type": "join",
+    ///             "next": { "builtin": "terminate" }
+    ///         }
+    ///     }
+    /// }
+    /// # "#)?;
+    /// # Ok::<_, serde_json::Error>(())
+    /// ```
+    Join(JoinOp),
+
+    /// If the request is serializable, transform it by running it through a [CEL](https://cel.dev/) program.
+    /// The context includes a "request" variable which contains the request.
+    ///
+    /// # Examples
+    /// ```
+    /// # bevy_impulse::Diagram::from_json_str(r#"
+    /// {
+    ///     "start": "transform",
+    ///     "ops": {
+    ///         "transform": {
+    ///             "type": "transform",
+    ///             "cel": "request.name",
+    ///             "next": { "builtin": "terminate" }
+    ///         }
+    ///     }
+    /// }
+    /// # "#)?;
+    /// # Ok::<_, serde_json::Error>(())
+    /// ```
+    ///
+    /// Note that due to how `serde_json` performs serialization, positive integers are always
+    /// serialized as unsigned. In CEL, You can't do an operation between unsigned and signed so
+    /// it is recommended to always perform explicit casts.
+    ///
+    /// # Examples
+    /// ```
+    /// # bevy_impulse::Diagram::from_json_str(r#"
+    /// {
+    ///     "start": "transform",
+    ///     "ops": {
+    ///         "transform": {
+    ///             "type": "transform",
+    ///             "cel": "int(request.score) * 3",
+    ///             "next": { "builtin": "terminate" }
+    ///         }
+    ///     }
+    /// }
+    /// # "#)?;
+    /// # Ok::<_, serde_json::Error>(())
+    /// ```
+    Transform(TransformOp),
+
+    /// Drop the request, equivalent to a no-op.
+    Dispose,
+}
+
+type DiagramStart = serde_json::Value;
+type DiagramTerminate = serde_json::Value;
+type DiagramScope<Streams = ()> = Scope<DiagramStart, DiagramTerminate, Streams>;
+
+#[derive(JsonSchema, Serialize, Deserialize)]
+#[serde(rename_all = "snake_case")]
+pub struct Diagram {
+    /// Version of the diagram, should always be `1`.
+    #[schemars(range(min = 1, max = 1))]
+    version: u32,
```

Comment:
I like the idea of a built-in version number, but I think we should use a [`SemVer`](https://github.com/open-rmf/rmf_site/blob/main/rmf_site_format/src/semver.rs#L38) so we can distinguish between forward-compatible and incompatible format changes.

I wrote that `SemVer` class several years ago so it may be worth checking if there's a small crate out in the ecosystem that implements the same concept.

Reply:
This crate may be worth considering: 

https://docs.rs/serde-semver/latest/serde_semver/

The only thing I don't like is it doesn't seem to let us inspect the actual version number that was present in the serialized data.

Reply:
9a231e726654ac73f9e9dfb7c5b50d89fee5f01a

Personally I don't think semver is needed, but it wouldn't hurt also so I changed it.